ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,943 @@
+
+| Field | Type     | Description                                 | Length   |
+|:------|:---------|:--------------------------------------------|:---------|
+| 1     | `varu64` | Message code: **must** have a value of `6`  | 1 byte   |
+| 2     | `coords` | Sender coordinates                          | Variable |
+| 3     | `bytes`  | Known bytes of target Node ID to search for | Variable |
+
+#### DHT response
+
+A DHT response is sent in response to a DHT request when a node on the DHT has
+knowledge that helps us to reach the target node. It **must** be encapsulated
+within a protocol message.
+
+A response may contain more than one set of target candidates, therefore the
+fields n<sub>1</sub> and n<sub>2</sub> may be repeated multiple times.
+
+| Field         | Type     | Description                                | Length   |
+|:--------------|:---------|:-------------------------------------------|:---------|
+| 1             | `varu64` | Message code: **must** have a value of `7` | 1 byte   |
+| 2             | `coords` | Sender coordinates                         | 8 bytes  |
+| 3             | `bytes`  | Known bytes of target Node ID searched for | 64 bytes |
+| n<sub>1</sub> | `bytes`  | Target candidate public encryption key     | 32 bytes |
+| n<sub>2</sub> | `coords` | Target candidate coordinates               | Variable |
+
+### Second-level link protocol message formats
+
+#### Switch update
+
+A switch update is a message that contains a fully signed path containing all
+hops from the root node down to a given node.
+
+The root node sends a switch update message to all directly connected peers,
+containing a timestamp and its own signing key. The switch update message
+generated by the root node contains no signatures.
+
+An update message that has been relayed by any non-root nodes will contain
+information about more than one hop, therefore the fields n<sub>1</sub> and
+n<sub>2</sub> will be repeated for each hop.
+
+When a node receives a switch update message, the node **must** append its own
+signing key to the update and then sign the newly-updated message. The
+newly-updated switch update message is then redistributed to all connected
+peers.
+
+This process continues until the switch update has flooded the entire network
+and all nodes have received a switch update that contains `n` number of
+signatures, where `n` is the number of hops from the node to the root.
+
+| Field         | Type     | Description                                | Length        |
+|:--------------|:---------|:-------------------------------------------|:--------------|
+| 1             | `varu64` | Message code: **must** have a value of `3` | 1 byte        |
+| 2             | `bytes`  | Root node public signing key               | 32 bytes      |
+| 3             | `varu64` | Timestamp                                  | 1 to 10 bytes |
+| n<sub>1</sub> | `bytes`  | Update node public signing key             | 32 bytes      |
```

Comment:
Before n<sub>1</sub>, there is a varu64 that is the port number of the node. I think this allows the connecting node to know the coordinates of itself and some other nodes.

Reply:
That varu64 also repeats for each update.

Reply:
Thanks for pointing out the error there.